### PR TITLE
Shallow snapshot v2 - create snapshot validations in a cluster state update

### PR DESCRIPTION
**Pull request by gbbafna**
### Description

1. Adding `IN_PROGRESS` cluster state entry for shallow v2 snapshots. Associated clean up with this. This ensure whatever validations we are doing are atomic in nature. 
2. Pinning timestamp before doing finalize snapshot .
3. Blocking concurrent v2 snapshots. 

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] Functionality includes testing.
- [ ] ~API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.~
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java [hunk 1]
```diff
@@ -468,26 +468,34 @@ public void createSnapshotV2(final CreateSnapshotRequest request, final ActionLi
         long pinnedTimestamp = System.currentTimeMillis();
         final String repositoryName = request.repository();
         final String snapshotName = indexNameExpressionResolver.resolveDateMathExpression(request.snapshot());
-        validate(repositoryName, snapshotName);
 
-        final SnapshotId snapshotId = new SnapshotId(snapshotName, UUIDs.randomBase64UUID()); // new UUID for the snapshot
         Repository repository = repositoriesService.repository(repositoryName);
+        repository.executeConsistentStateUpdate(repositoryData -> new ClusterStateUpdateTask(Priority.URGENT) {
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:473]: Why is priority `URGENT`? For non-v2 snapshots, we use `NORMAL` priority.

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java [hunk 2]
```diff
@@ -509,74 +517,130 @@ public void createSnapshotV2(final CreateSnapshotRequest request, final ActionLi
...
+                    threadPool.absoluteTimeInMillis(),
+                    repositoryData.getGenId(),
+                    shards,
+                    userMeta,
+                    version,
+                    true,
+                    true
+                );
+                final List<SnapshotsInProgress.Entry> newEntries = new ArrayList<>(runningSnapshots);
+                newEntries.add(newEntry);
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:543]: When do we remove this entry from cluster state? On snapshot completion?

File: server/src/main/java/org/opensearch/cluster/SnapshotsInProgress.java [hunk 3]
```diff
@@ -134,6 +134,38 @@ public static Entry startedEntry(
...
+        List<IndexId> indices,
+        List<String> dataStreams,
+        long startTime,
+        long repositoryStateId,
+        final Map<ShardId, ShardSnapshotStatus> shards,
+        Map<String, Object> userMetadata,
+        Version version,
+        boolean remoteStoreIndexShallowCopy,
+        boolean remoteStoreIndexShallowCopyV2
+    ) {
```
[the remark this answers is not in the document]
Comment (gbbafna) [reply server/src/main/java/org/opensearch/cluster/SnapshotsInProgress.java:150]: It is doing the same as of now .

File: server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java [hunk 4]
```diff
@@ -3392,6 +3396,7 @@ public ClusterState execute(ClusterState currentState) {
                                     + "]"
                             );
                         }
+                        logger.info("Done Setting it to {} {}", newGen, newGen);
```
[the remark this answers is not in the document]
Comment (gbbafna) [reply server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java:3399]: Yes, will remove these lines .

File: server/src/main/java/org/opensearch/cluster/SnapshotsInProgress.java [hunk 5]
```diff
@@ -243,11 +344,12 @@ public Entry(
...
                 null,
                 Map.of(),
-                remoteStoreIndexShallowCopy
+                remoteStoreIndexShallowCopy,
+                false
             );
         }
 
-        private Entry(
+        public Entry(
```
[the remark this answers is not in the document]
Comment (gbbafna) [reply server/src/main/java/org/opensearch/cluster/SnapshotsInProgress.java:352]: Good catch. no, removed it .

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java [hunk 6]
```diff
@@ -468,26 +468,34 @@ public void createSnapshotV2(final CreateSnapshotRequest request, final ActionLi
...
-            final StepListener<RepositoryData> repositoryDataListener = new StepListener<>();
-            repositoriesService.getRepositoryData(repositoryName, repositoryDataListener);
+            private Snapshot snapshot;
+
+            boolean enteredLoop;
+
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                // move to in progress
+                validate(repositoryName, snapshotName);
```
[the remark this answers is not in the document]
Comment (gbbafna) [reply server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:485]: Agreed , moved it before cluster state update now .

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java (hunk 1 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:473]: This is v2 snapshots only .

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java (hunk 2 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:543]: Yes . See line no 604 . 

```
state -> stateWithoutSnapshot(state, snapshot),
```

**Review by sachinpkale — APPROVED** [review]

**Review by sachinpkale — APPROVED** [review]

**Pull request merged**